infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/project/OpenShiftPods.java

```diff
@@ -381,28 +389,38 @@ public void delete() throws InfrastructureException {
       final CompletableFuture<Void> removed =
           allOf(deleteFutures.toArray(new CompletableFuture[deleteFutures.size()]));
       try {
-        removed.get();
+        removed.get(POD_REMOVAL_TIMEOUT_MIN, TimeUnit.MINUTES);
       } catch (InterruptedException e) {
         Thread.currentThread().interrupt();
         throw new InfrastructureException(
             "Interrupted while waiting for pod removal. " + e.getMessage());
       } catch (ExecutionException e) {
         throw new InfrastructureException(
             "Error occurred while waiting for pod removing. " + e.getMessage());
+      } catch (TimeoutException ex) {
+        throw new InfrastructureException("Pods removal timeout reached " + ex.getMessage());
```

Comment:
Can you elaborate on what will happen in this situation with cleanup of other resources and with the possible following start of the same workspace?

Reply:
It is similar to other exception that is possible on cleanup, I would say that we should revise cleanup in general. Because if any step of cleanup is failed then, other resources won't be cleaned. I'll create an issue for that.

Reply:
https://github.com/eclipse/che/issues/7660